### ISSUE TITLE
Dist files

### ISSUE THREAD

**Issue by 1j01** [hollowdoor/dom_autoscroller#12]
[dom-autoscroller.js](https://github.com/hollowdoor/dom_autoscroller/blob/b8b1ab00092587191266006733ef797acdafbea1/dist/dom-autoscroller.js) references `createPointCB` which is undefined.
[dom-autoscroller.min.js](https://github.com/hollowdoor/dom_autoscroller/blob/b8b1ab00092587191266006733ef797acdafbea1/dist/dom-autoscroller.min.js) is just `[object Object]`

Ideally the npm module and browser bundle should be the same and should work in any environment, so that you could for instance use `<script src="https://unpkg.com/dom-autoscroller@2.0.0">`

**Comment by hollowdoor** [comment]
Ah crap! That was stupid. It was a change in the build.

It's fixed now. As 2.1.0

**Comment by hollowdoor** [comment]
Also are you using the dist build, or are you hanging around helping me?

**Comment by hollowdoor** [comment]
Interesting. I've never heard of unpkg.com before.

For reference here's the link to the complete links for this.

https://unpkg.com/dom-autoscroller@2.1.1/dist/dom-autoscroller.js

https://unpkg.com/dom-autoscroller@2.1.1/dist/dom-autoscroller.min.js

Also I forgot to fix the minified build. Fixed as of 2.1.1. What a mess.

**Comment by 1j01** [comment]
I was going to answer a [question on SO](http://stackoverflow.com/questions/17895141/how-to-scroll-the-window-automatically-when-mouse-moves-bottom-of-the-page-using) with a jsfiddle using the module. ([Still not quite working.](https://github.com/hollowdoor/dom_autoscroller/issues/13))
If I were to use this module in my app it would be through a bundler. To to that though I think I'd need a hook to run some code in the animation loop though, need to update the selection smoothly as it scrolls.

**Comment by hollowdoor** [comment]
What do you mean by a hook?

Do you mean some way to share some other animation (selection) with the autoscroll animation.

I think that could be possible but I hesitate to make it consume something. Maybe an updating method could be exported to be shared with other animations like:

``` javascript
requestAnimationFrame(()=>{
    scroller.update();
});
```

or

``` javascript
let update = scroller.callback();
requestAnimationFrame(()=>{
    update();
});
```

The above style would mitigate dependency lock-in.

That would take quite a bit of thought, and would be a little weird though. With the scroller being based on mouse/touch move the startup might be a little hard to work out.

**Comment by 1j01** [comment]
I think it can just be a callback that you pass in as an option (like `scrollHook`, `onScroll`, or `onUpdate`).
I also might be able to just use a separate animation loop, but it might not have the same timing characteristics in relation to the scroll offset updating or might just not be as nice (i.e. I would have to start and stop a `requestAnimationFrame` loop in parallel with dom-autoscroller)
I'll try and replace my custom solution with dom-autoscroller next time I work on that project, and I'll let you know how it goes.
([Here's the code that would go in the callback, by the way](https://github.com/1j01/wavey/blob/d15915cbf99f0129e7d06690e9bed54d856002de/src/components/TracksArea.coffee#L191-L194), [copied from mousemove](https://github.com/1j01/wavey/blob/d15915cbf99f0129e7d06690e9bed54d856002de/src/components/TracksArea.coffee#L212-L214) (TODO: DRY))

**Comment by hollowdoor** [comment]
Would a listener like this be workable for you?

``` javascript
function onUpdateListener(coords={x, y}){

}
```

Even if I do this in the long run I'd rather "get" a func than "put" a listener.

This conversation made me think of my own future work. I'm familiar with game loops. So it wouldn't be too confusing to restructure based on a game loop concept. Maybe not as easy as an update listener. ;)

I can do both. I see how someone would prefer a listener. I'm also thinking an external loop might be very beneficial as well. Like if someone is using TweenJS, and dom-autoscroller in the same app.

> (i.e. I would have to start and stop a requestAnimationFrame loop in parallel with dom-autoscroller)

Maybe. I'm thinking it would be a full fledged game loop. Stopping, and starting would be done with property logic specific to separate objects.

Like this:

With two objects. Object one has an update, and render methods. Object two has an update, and render methods.
1. Object one, and two update themselves, and expose render methods.
2. A requestAnimationFrame loop constantly runs calling both Object's render methods every loop.
3. Internal logic to Object one, and two controls actual rendering based on what happens inside their respective internal update methods.

I've gotta say you kind of inspired me. I looked at your sound editor app wavey. I'm starting to get some ideas about how libraries could work in an environment like that. :)

**Comment by 1j01** [comment]
I'm guessing you're thinking of sharing `requestAnimationFrame` loops for a performance benefit, but I don't think it would be significant if even measurable, so that would probably all be premature optimization.
For instance in my app I actually have another `requestAnimationFrame` loop that's always going (that hopefully I can get rid of at some point) and it doesn't interfere with autoscrolling or anything.

I was going to say yes, the update function with coordinates as arguments should be good, but I realized that's not actually the right solution. The problem is that objects being dragged (in my case a selection) only update on `mousemove` so don't get updated in sync with the scrolling. In my case I could use the callback to update the selection, but libraries like jquery.nestable don't provide a way to manually update the elements being dragged, so you would need to simulate a `mousemove` event.
I also realized I introduced bugs by copying over only some of the code from my `mousemove` handler, because really it should be done only from `mousemove`.
Ideally browsers would emit `mousemove` events when a container was scrolled by JS, but in lieu of that,
and in conclusion, [dom-autoscroller should trigger `mousemove` events](https://github.com/hollowdoor/dom_autoscroller/issues/14).
(Opened a new issue because we're completely off-topic at this point.)

**Comment by hollowdoor** [comment]
dom-autoscroller should **not** trigger mouse move events. :)

You didn't introduce bugs. I can interpret the code with my own thinking. So you're blameless.

> so that would probably all be premature optimization

My desktop is a worn out fan cooled duel core from 2011 that crashes when watching 720p video. I feel every looping animation, and mouse move event as sweaty palms.

That being said. My main concern was doing I'll call you instead of you call me. You in this case being dom-autoscroller. The optimization is more of a side benefit.

**Comment by 1j01** [comment]
Sorry if I was unclear, I introduced bugs in my _application code_, by not reusing enough of the `mousemove` event logic. Specifically the selection doesn't start updating while autoscrolling if you start dragging from within the scrolling margin and move the mouse very little. It's very much an edge-case bug. I should use [all of this logic](https://github.com/1j01/wavey/blob/d15915cbf99f0129e7d06690e9bed54d856002de/src/components/TracksArea.coffee#L207-L215) in both places, but since I was copying and pasting I didn't want to make it too WET so I copied as little as I could that still seemed to work.

I don't think you'll see any performance benefit from consolidating rAF loops. I think the performance difference would be on the order of calling a function that calls two functions versus calling two functions. The actual stuff that the functions do, such as with TweenJS updating animations, and with dom-autoscroller scrolling containers and triggering reflows, both triggering repaints, would be orders of magnitude more time-consuming. That's the stuff you really have to worry about.

**Comment by hollowdoor** [comment]
Like I said my main concern isn't about performance. I'm concerned about -- I'll call dom-autoscroller instead of dom-autoscroller calling me. I want inversion of control to go to the user code instead of to dom-autoscroller. 

That's my preference. There isn't much definitive on what's best for any use case since it's just me, and you arguing in here. Like the problem with nestable. I never know when there will be some conflict with someone's usage that conflicts with someone else's usage. In the case of the inception of dom-autoscroller I started with compatibility with dragula because I like it's independence.

Personally I find jQuery/Zepto plugins annoying because of that hard dependency where control is only allowed through options which is why we're in this mess #14  . If nestable didn't use jQuery, and they allowed more external control of that library I feel things would be much better. Which is where I'd like to take dom-autoscroller.

**Comment by 1j01** [comment]
I'm not disagreeing about the inversion of control, but I don't understand what you're getting at.
Wouldn't a callback (or the triggering of an event) _be_ inversion of control?
(I'm also not arguing for jQuery - don't generally use it myself)

**Comment by hollowdoor** [comment]
An event would be inversion of control. I view it as the choice between two consumers of dependencies. The library can consume the dependency, or the user code can consume the dependency.

In fact that's exactly how it works in libraries like Three.js, and TweenJS. They have an update method that _you_ call.

There's a similar concept in the Mediator pattern. The Mediator calls all the other stuff, or the others call Mediator. The Mediator acts as a host object. Of course it wouldn't be exactly the same inside a loop because it would be a continuous repeated call, but I hope you get what I mean.
